samza-core/src/test/java/org/apache/samza/runtime/TestRemoteApplicationRunner.java

```diff
@@ -75,9 +75,20 @@ public void testWaitForFinishTimesout() {
     assertFalse("Application finished before the timeout.", finished);
   }
 
+  @Test
+  public void testRunWithConfigLoaderFactoryPresent() {
```

Comment:
Haha, I figured that too. In general, I prefer not to mess with production code for the sake of unit tests.

The current unit actually test the new flow, I passed null SamzaApplication to RemoteApplicationRunner's constructor. In RemoteApplicationRunner#run, if it goes the legacy flow, it will throw exception when constructing RemoteJobPlanner where SamzaApplication is required. Do you think this is enough?

Reply:
There is some precedence elsewhere in samza code; ZkJobCoordinator.readJobModelFromMetadataStore for example uses a similar method structure + spy in unit test.

You are correct that the test written now does ensure that the legacy flow isn't triggered, but it does little to verify the behavior of the new flow. I won't block the PR on this though.

Reply:
+1 that there is a bit of missing test coverage here. It would be nice if this class was set up better for improved testing, but I also won't block the PR on this.